skimage/util/_backends.py

```diff
@@ -0,0 +1,153 @@
+        return True
+    else:
+        return False
+
+
+def get_module_name(func):
+    """Get the public module name of a scikit-image function.
+
+    This computes the name of the public submodule in which the function can
+    be found.
+    """
+    full_name = func.__module__
+    # This relies on the fact that scikit-image does not use
+    # sub-submodules in its public API, except in one case.
+    # This means that public name can be atmost `skimage.foobar`
+    # for everything else
+    if full_name.startswith("skimage.filters.rank"):
+        public_name = "skimage.filters.rank"
+    else:
+        public_name = ".".join(full_name.split(".")[:2])
+
+    # It would be nice to sanity check things by doing something like the
+    # following. However we can't because this code is executed while the
+    # module is being imported, which means this would create a circular
+    # import
+    # mod = importlib.import_module(public_name)
+    # assert getattr(mod, func.__name__) is func
+
+    return public_name
+
+
+@cache
+def all_backends():
+    """List all installed backends and information about them"""
+    backends = {}
+    backends_ = _entry_points(group="skimage_backends")
+    backend_infos = _entry_points(group="skimage_backend_infos")
+
+    for backend in backends_:
+        backends[backend.name] = {"implementation": backend}
+        try:
+            info = backend_infos[backend.name]
+            # Double () to load and then call the backend information function
+            backends[backend.name]["info"] = info.load()()
+        except KeyError:
+            pass
+
+    return backends
+
+
+def dispatchable(func):
+    """Mark a function as dispatchable.
+
+    When a decorated function is called the installed backends are
+    searched for an implementation. If no backend implements the function
+    then the scikit-image implementation is used.
+    """
+    func_name = func.__name__
+    func_module = get_module_name(func)
+
+    # If no backends are installed at all or dispatching is disabled,
+    # return the original function. This way people who don't care about it
+    # don't see anything related to dispatching
+    if disable_dispatching() or not all_backends():
+        return func
+
+    @functools.wraps(func)
+    def wrapper(*args, **kwargs):
+        # Backends are tried in alphabetical order, this makes things
+        # predictable and stable across runs. Might need a better solution
+        # when it becomes common that users have more than one backend
+        # that would accept a call.
+        for name in sorted(all_backends()):
+            backend = all_backends()[name]
+            # If we have a BackendInformation object we check if the
+            # function we are looking for is implemented in the backend
+            if "info" in backend:
+                if (
+                    f"{func_module}:{func_name}"
+                    not in backend["info"].supported_functions
```

Comment:
Could we just merge this functionality into `can_has`? That way the interface may get even simpler? I don't remember if there was an intentional reason for doing it this way.

Reply:
We could remove this and just rely on `can_has`.

There were two reasons why having this would be useful: (1) docs building and (2) avoid expensive imports/init

For (1) I think my original thinking was that having a list of all functions that a backend implements would be useful (and needed) for being able to highlight/mark functions in the docs as "has alternative backend". You can't really use `can_has` for that use case as the return value of that can depend on the values of the arguments.

The question is, do we need this list of functions? Maybe we don't because a backend publishes a JSON document that is loaded dynamically when browsing the docs. And really, if we go with the "JSON doc" approach then having a list like `supported_functions` isn't great. It is just going to be out of date/different from the JSON doc's list.

For (2) I think it is still useful. That is because we have the contract that the info entrypoint has to be importable without any special hardware or software being present. For example, you should be able to use the info entrypoint of "Tim's special backend" even if you don't have my fancy proprietary hardware.

I also think that implementing a `can_has` that can perform the equivalent action without importing libraries that are expensive to initialise (e.g. CUDA stuff) and at the same time can check the types of the arguments or some such (for which you'd want your CUDA library imported presumably) is hard. So by having this split between the info part which is literally a list of strings (that you could compute with a separate script and hardcode in your backend) and a `can_has` that is a fully fledged Python function we help backend implementers to not shoot themselves in the foot.